### PR TITLE
[Experiment] feat(17227): Support `@abstract` JSDoc tag

### LINKED ISSUE #17227: Support `@abstract` JSDoc tag
**TypeScript Version:**  nightly (2.5.0-dev.20170712)

**Code**

```ts
/** @abstract */
class C {}
new C();
```

**Expected behavior:**

Error.

**Actual behavior:**

No error.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #17227

### CODE REVIEW

File: tests/cases/conformance/jsdoc/jsdocAbstract2.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// @allowJs: true
+// @checkJs: true
+// @target: esnext
+// @noEmit: true
```
Comment (ajafff) [inline tests/cases/conformance/jsdoc/jsdocAbstract2.ts:4]: can you change this to `@emitDeclarationOnly: true` to also test the declaration emit?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -31899,7 +31900,7 @@ namespace ts {
 
             // Abstract methods cannot have an implementation.
             // Extra checks are to avoid reporting multiple errors relating to the "abstractness" of the node.
-            if (hasSyntacticModifier(node, ModifierFlags.Abstract) && node.kind === SyntaxKind.MethodDeclaration && node.body) {
+            if (hasSyntacticModifier(node, ModifierFlags.Abstract) && node.kind === SyntaxKind.MethodDeclaration && node.body && !isInJSFile(node)) {
```
Comment (ajafff) [inline src/compiler/checker.ts:31903]: I wondered why this change is necessary as there cannot be syntactic modifiers in JS files. In case you also wondered: it turns out this might be a bug: #42189

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -33277,6 +33278,13 @@ namespace ts {
             checkSignatureDeclaration(node);
         }
 
+        function checkJSDocAbstractTag(node: JSDocAbstractTag): void {
+            const host = getEffectiveJSDocHost(node);
+            if (host && isClassElement(host) && !hasEffectiveModifier(host.parent, ModifierFlags.Abstract)) {
+                error(host, Diagnostics.Abstract_methods_can_only_appear_within_an_abstract_class);
```
Comment (ajafff) [inline src/compiler/checker.ts:33284]: properties can be abstract too. the error message doesn't fit in that case

File: tests/baselines/reference/jsdocAbstract2.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    declare abstract class A {
+        /**
+         * @abstract
+         * @returns {number}
+         */
+        abstract foo(): number;
+    }
+    declare class B extends A {
+                  ~
+!!! error TS2515: Non-abstract class 'B' does not implement inherited abstract member 'foo' from class 'A'.
```
Comment (ajafff) [inline tests/baselines/reference/jsdocAbstract2.js:50]: This error should not be there. The of `class B` output is missing `foo(): number`

File: tests/cases/conformance/jsdoc/jsdocAbstract2.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+     * @returns {number}
+     */
+    foo() {}
+}
+
+class B extends A {
+    foo() {
+        return 1;
+    }
+}
```
Comment (ajafff) [inline tests/cases/conformance/jsdoc/jsdocAbstract2.ts:21]: can you add a few more test cases:

```ts
class C extends A {} // expects error that method 'foo' is not implemented

/** @abstract */
class D extends A {} // no error
```

File: tests/cases/conformance/jsdoc/jsdocAbstract4.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+// @allowJs: true
+// @checkJs: true
+// @target: esnext
+// @declaration: true
+// @emitDeclarationOnly: true
+// @Filename: jsdocAbstract4.js
+
+class Foo {
+    /** @abstract */
+    foo() {}
```
Comment (ajafff) [inline tests/cases/conformance/jsdoc/jsdocAbstract4.ts:10]: can you add more abstract class elements: propery, get and set accessor

Comment (ajafff) [inline tests/cases/conformance/jsdoc/jsdocAbstract4.ts:10]: consider adding a test with a static abstract method to ensure there is an error either on `static` or `@abstract`

File: src/compiler/checker.ts (hunk 3 above)
Comment (ajafff) [reply src/compiler/checker.ts:33284]: nevermind, in TS the error text is the same for properties

File: src/compiler/checker.ts (hunk 2 above)
Comment (a-tarasyuk) [reply src/compiler/checker.ts:31903]: `hasSyntacticModifier` does not include `JSDoc`.

File: tests/cases/conformance/jsdoc/jsdocAbstract4.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+    /** @abstract */
+    p1;
+
+    /** @abstract */
+    get p2() {}
+
+    /** @abstract */
+    set p2() {}
+}
```
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply tests/cases/conformance/jsdoc/jsdocAbstract4.ts:23]: I see many comments for this pr, so I suppose someone can make it better.

[the remark this answers is not in the document]
Comment (ajafff) [reply tests/cases/conformance/jsdoc/jsdocAbstract4.ts:23]: I'm sorry for the amount of feedback. I didn't mean to criticise your great work. I only wanted to make sure that edge cases are considered.

Most of the test cases above are probably already covered by your implementation. Adding a test for them simply makes sure it doesn't regress in the future.

**Comment by ajafff** [comment]
We also need to determine whether legacy class syntax should be recognized:

```js
/**
 * @class
 * @abstract
 */
function Legacy() {
  /**
   * @type {string}
   * @abstract
   */
  this.prop;
}

/** @abstract */
Legacy.prototype.doStuff = function() {}
```

**Review by sandersn — CHANGES_REQUESTED** [review]
Here's what I think is needed to finish this feature:

1. Survey existing code to see whether requiring `@abstract` methods only to occur in `@abstract` classes would result in lots of errors.
2. Figure out how serialisation code results in issuing errors, then move the error earlier.
3. Add test cases.
4. Restrict what's allowed in the body of abstract methods.

File: tests/baselines/reference/jsdocAbstract2.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+/** @abstract */
+declare abstract class A {
+    /**
+     * @abstract
+     * @returns {number}
+     */
+    abstract foo(): number;
+}
+declare class B extends A {
+}
```
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply tests/baselines/reference/jsdocAbstract2.js:61]: @sandersn

```ts
abstract class A {
    abstract foo(): number;
}

class B extends A {
    foo() {
        return 1;
    }
}

// .d.ts
declare abstract class A {
    abstract foo(): number;
}
declare class B extends A {
    foo(): number;
}
```

[Playground](https://www.typescriptlang.org/play?ssl=9&ssc=2&pln=1&pc=1#code/IYIwzgLgTsDGEAJYBthjAgggg3gWACgFiFRIZ4EAzAexoAoBKALgQDsBXAWxAFMoA3IQC+hQijQYAQgl4APCLzYATDNnxEStBo1yESBhFF4QOUNggCMQzcVEFhQA)


```js
/** @abstract */
class A {
    /**
     * @abstract
     * @returns {number}
     */
    foo() { }
}
// ok
class B extends A {
    foo() {
        return 1;
    }
}

// .d.ts
/** @abstract */
declare abstract class A {
    /**
     * @abstract
     * @returns {number}
     */
    abstract foo(): number;
}
declare class B extends A {
    foo(): number;
}
```

File: src/compiler/checker.ts [hunk 9]
```diff
@@ -33277,6 +33292,15 @@ namespace ts {
             checkSignatureDeclaration(node);
         }
 
+        function checkJSDocAbstractTag(node: JSDocAbstractTag): void {
+            const host = getEffectiveJSDocHost(node);
+            if (host && isClassElement(host)) {
+                if (!hasEffectiveModifier(host.parent, ModifierFlags.Abstract)) {
+                    error(host, Diagnostics.Abstract_methods_can_only_appear_within_an_abstract_class);
```
[the remark this answers is not in the document]
Comment (sandersn) [reply src/compiler/checker.ts:33299]: Based on [my analysis](https://github.com/microsoft/TypeScript/issues/17227#issuecomment-758154771) on the original issue the majority of existing usages will have an error here.

[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply src/compiler/checker.ts:33299]: @sandersn ok, I'll remove it.

**Comment by sandersn** [comment]
Here are a couple of observations based on [my survey of existing usage](https://github.com/microsoft/TypeScript/issues/17227#issuecomment-758154771):

1. The majority of existing `@abstract` on methods will get an error from the code currently in `checkJSDocAbstractTag`.
2. There's not much usage, though, so it wouldn't inconvenience many people.
3. Because there's so little usage already, I'm not sure this feature is worth it.

To accept this I think we'd need buyin from an existing user who wants this feature and agrees to test that it meets their needs. Looking at usage rates, only webpack or discord.js make sense to me.

File: src/compiler/checker.ts [hunk 10]
```diff
@@ -39851,7 +39877,7 @@ namespace ts {
                     return grammarErrorAtPos(accessor, accessor.end - 1, ";".length, Diagnostics._0_expected, "{");
                 }
             }
-            if (accessor.body && hasSyntacticModifier(accessor, ModifierFlags.Abstract)) {
+            if (accessor.body && hasEffectiveModifier(accessor, ModifierFlags.Abstract) && !isInJSFile(accessor)) {
```
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply src/compiler/checker.ts]: @sandersn Done

File: tests/baselines/reference/jsdocAbstract4.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    get p2() { }
+    /** @abstract */
+    set p2() { }
+}
+
+
+//// [jsdocAbstract4.d.ts]
+declare class Foo {
+    /** @abstract */
+    abstract static m2(): void;
```
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply tests/baselines/reference/jsdocAbstract4.js]: `abstract static` illegal in TS files. Do we need to throw an error in JS files like in TS? Or need to skip emitting `.d.ts` if there is an error?

```ts
// foo.ts
abstract class Foo {
    abstract static foo = 1;
             ^^^^^ TS1243: 'static' modifier cannot be used with 'abstract' modifier.
}

// foo.d.ts
declare abstract class Foo {
    abstract static foo: number;
}
```

**Comment by sandersn** [comment]
@typescript-bot pack this

**Comment by sandersn** [comment]
@typescript-bot pack this

**Comment by peter50216** [comment]
@typescript-bot pack this

**Comment by andrewbranch** [comment]
@typescript-bot pack this

**Comment by trusktr** [comment]
@a-tarasyuk Can you please restore the content of the PR so that someone else can give it a shot?

**Comment by trusktr** [comment]
Is there some way to clone/pull the deleted commit, https://github.com/microsoft/TypeScript/commit/dfeab9e4522e64e17174cae3ef700b410f6e4352. It appears to be visible on GitHub, but there's a note saying it does not exist in the repo. Same with https://github.com/a-tarasyuk/TypeScript/commit/dfeab9e4522e64e17174cae3ef700b410f6e4352.

**Comment by forivall** [comment]
@trusktr idk you're still wondering, but it can be fetched by `git fetch origin dfeab9e4522e64e17174cae3ef700b410f6e4352 && git branch feat/17227 FETCH_HEAD` - this will restore the branch into your local checkout.
